core/ibft.go

```diff
@@ -0,0 +1,1134 @@
+}
+
+// moveToNewRound moves the state to the new round
+func (i *IBFT) moveToNewRound(round uint64) {
+	i.state.setView(&proto.View{
+		Height: i.state.getHeight(),
+		Round:  round,
+	})
+
+	i.state.setRoundStarted(false)
+	i.state.setProposalMessage(nil)
+	i.state.changeState(newRound)
+}
+
+func (i *IBFT) buildProposal(ctx context.Context, view *proto.View) *proto.Message {
+	var (
+		height = view.Height
+		round  = view.Round
+	)
+
+	if round == 0 {
+		proposal := i.backend.BuildProposal(height)
+
+		return i.backend.BuildPrePrepareMessage(
+			proposal,
+			nil,
+			&proto.View{
+				Height: height,
+				Round:  round,
+			},
+		)
+	}
+
+	//	round > 0 -> needs RCC
+	rcc := i.waitForRCC(ctx, height, round)
+	if rcc == nil {
+		// Timeout occurred
+		return nil
+	}
+
+	//	check the messages for any previous proposal (if they have any, it's the same proposal)
+	var previousProposal []byte
+
+	for _, msg := range rcc.RoundChangeMessages {
+		//	if message contains block, break
+		latestPC := messages.ExtractLatestPC(msg)
+
+		if latestPC != nil {
+			previousProposal = messages.ExtractLastPreparedProposedBlock(msg)
+
+			break
+		}
+	}
+
+	if previousProposal == nil {
+		//	build new proposal
+		proposal := i.backend.BuildProposal(height)
+
+		return i.backend.BuildPrePrepareMessage(
+			proposal,
+			rcc,
+			&proto.View{
+				Height: height,
+				Round:  round,
+			},
+		)
+	}
+
+	return i.backend.BuildPrePrepareMessage(
+		previousProposal,
+		rcc,
+		&proto.View{
+			Height: height,
+			Round:  round,
+		},
+	)
+}
+
+// acceptProposal accepts the proposal and moves the state
+func (i *IBFT) acceptProposal(proposalMessage *proto.Message) {
+	//	accept newly proposed block and move to PREPARE state
+	i.state.setProposalMessage(proposalMessage)
+	i.state.changeState(prepare)
+}
+
+// AddMessage adds a new message to the IBFT message system
+func (i *IBFT) AddMessage(message *proto.Message) {
+	// Make sure the message is present
+	if message == nil {
+		return
+	}
+
+	// Check if the message should even be considered
+	if i.isAcceptableMessage(message) {
+		i.messages.AddMessage(message)
+	}
+}
+
+// isAcceptableMessage checks if the message can even be accepted
+func (i *IBFT) isAcceptableMessage(message *proto.Message) bool {
+	//	Make sure the message sender is ok
+	if !i.backend.IsValidSender(message) {
+		return false
+	}
+
+	// Invalid messages are discarded
+	if message.View == nil {
+		return false
+	}
+
+	// Make sure the message is in accordance with
+	// the current state height, or greater
+	if i.state.getHeight() > message.View.Height {
+		return false
+	}
+
+	// Make sure the message round is >= the current state round
+	return message.View.Round >= i.state.getRound()
+}
+
+//	ExtendRoundTimeout extends each round's timer by the specified amount.
+func (i *IBFT) ExtendRoundTimeout(amount time.Duration) {
+	i.additionalTimeout = amount
+}
+
+// validPC verifies that  the prepared certificate is valid
+func (i *IBFT) validPC(
```

Comment:
Just thought it's hard to distinguish between `RC` and `PC` 😭

Reply:
It was a nightmare to get straight when reading the spec 🙃

Reply:
The key thing to remember is that RCC (Round Change Certificate) is just a collection of Round Change messages, whereas the PC (Prepared Certificate) is a struct that has specific fields